File: packages/cli/hieudoanm/cobra.go/src/cmd/search/text/text.go
package text

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/spf13/cobra"
)

type textMatch struct {
	File    string `json:"file"`
	Line    int    `json:"line"`
	Content string `json:"content"`
}

func includeToRegex(include string) *regexp.Regexp {
	if include == "" {
		return nil
	}
	globRe := strings.ReplaceAll(regexp.QuoteMeta(include), "\\*", ".*")
	globRe = "^" + globRe + "$"
	return regexp.MustCompile(globRe)
}

func searchTextInRoots(pattern string, searchRoots []string, re *regexp.Regexp, include *regexp.Regexp, maxDepth, maxCount int) ([]textMatch, error) {
	var results []textMatch
	seen := make(map[string]bool)

	for _, root := range searchRoots {
		info, err := os.Stat(root)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error: %v\n", err)
			continue
		}

		if !info.IsDir() {
			matches := searchFileText(root, pattern, re, maxCount)
			for _, m := range matches {
				key := m.File + ":" + m.Content
				if seen[key] {
					continue
				}
				seen[key] = true
				results = append(results, m)
			}
			continue
		}

		walkTextFile(re, root, maxDepth, include, &results, maxCount, seen)
	}
	return results, nil
}

func outputTextResults(results []textMatch, pattern string, jsonOutput bool) error {
	if jsonOutput {
		out, err := json.MarshalIndent(map[string]interface{}{
			"pattern": pattern,
			"matches": len(results),
			"results": results,
		}, "", "  ")
		if err != nil {
			return err
		}
		fmt.Println(string(out))
		return nil
	}

	if len(results) == 0 {
		fmt.Println("(no matches)")
		return nil
	}

	for _, m := range results {
		fmt.Printf("%s:%d: %s\n", m.File, m.Line, m.Content)
	}
	if len(results) > 1 {
		fmt.Printf("\n%d matches\n", len(results))
	}
	return nil
}

func searchFileText(path, pattern string, re *regexp.Regexp, maxCount int) []textMatch {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var isBinary bool
	for _, b := range data {
		if b == 0 {
			isBinary = true
			break
		}
	}
	if isBinary {
		return nil
	}

	lines := strings.Split(string(data), "\n")
	var matches []textMatch
	for i, line := range lines {
		if re.MatchString(line) {
			matches = append(matches, textMatch{
				File:    path,
				Line:    i + 1,
				Content: line,
			})
			if maxCount > 0 && len(matches) >= maxCount {
				break
			}
		}
	}
	return matches
}

func NewCommand() *cobra.Command {
	var pattern string
	var path string
	var ignoreCase bool
	var maxCount int
	var include string
	var maxDepth int

	cmd := &cobra.Command{
		Use:   "text [--pattern <pattern>] [--path <path>]",
		Short: "Search file contents using regex",
		Long:  `Search for a regex pattern inside files. If a directory is given, searches recursively.`,
		Example: `  search text --pattern "TODO" --path .
  search text --pattern "func.*error" --include "*.go"
  search text --pattern "import" --path src/ --ignore-case
  search text --pattern "panic" --max-count 5`,
		RunE: func(cmd *cobra.Command, args []string) error {
			jsonOutput, _ := cmd.Flags().GetBool("json")

			rePattern := pattern
			if ignoreCase {
				rePattern = "(?i)" + pattern
			}

			re, err := regexp.Compile(rePattern)
			if err != nil {
				return fmt.Errorf("invalid regex %q: %w", pattern, err)
			}

			includePattern := includeToRegex(include)
			results, err := searchTextInRoots(pattern, []string{path}, re, includePattern, maxDepth, maxCount)
			if err != nil {
				return err
			}

			return outputTextResults(results, pattern, jsonOutput)
		},
	}

	cmd.Flags().StringVarP(&pattern, "pattern", "p", "", "Regex pattern to search")
	cmd.Flags().StringVarP(&path, "path", "P", ".", "File or directory to search")
	cmd.Flags().BoolVarP(&ignoreCase, "ignore-case", "i", false, "Case-insensitive search")
	cmd.Flags().IntVarP(&maxCount, "max-count", "m", 0, "Maximum number of matches")
	cmd.Flags().StringVar(&include, "include", "", "Glob pattern for file names (e.g. \"*.go\")")
	cmd.Flags().IntVarP(&maxDepth, "max-depth", "d", 0, "Maximum directory depth (0 = unlimited)")
	return cmd
}

File: packages/cli/jack/clap.rs/src/cmd/search/text.rs
use anyhow::Context;
use serde::Serialize;

#[derive(Serialize)]
struct TextMatch {
    file: String,
    line: usize,
    content: String,
}

#[derive(clap::Args)]
pub struct Args {
    #[arg(short = 'p', long = "pattern", help = "Regex pattern to search")]
    pub pattern: String,
    #[arg(
        short = 'P',
        long = "path",
        default_value = ".",
        help = "File or directory to search"
    )]
    pub path: String,
    #[arg(short = 'i', long = "ignore-case", action = clap::ArgAction::SetTrue, help = "Case-insensitive search")]
    pub ignore_case: bool,
    #[arg(
        short = 'm',
        long = "max-count",
        default_value = "0",
        help = "Maximum number of matches"
    )]
    pub max_count: String,
    #[arg(long = "include")]
    pub include: Option<String>,
    #[arg(
        short = 'd',
        long = "max-depth",
        default_value = "0",
        help = "Maximum directory depth (0 = unlimited)"
    )]
    pub max_depth: String,
    #[arg(long = "json", action = clap::ArgAction::SetTrue, help = "Output in JSON format")]
    pub json: bool,
}

pub fn command() -> clap::Command {
    clap::Command::new("text")
        .about("Search file contents using regex")
        .arg(
            clap::Arg::new("pattern")
                .long("pattern")
                .short('p')
                .help("Regex pattern to search")
                .required(true),
        )
        .arg(
            clap::Arg::new("path")
                .long("path")
                .short('P')
                .help("File or directory to search")
                .default_value("."),
        )
        .arg(
            clap::Arg::new("ignore_case")
                .long("ignore-case")
                .short('i')
                .action(clap::ArgAction::SetTrue)
                .help("Case-insensitive search"),
        )
        .arg(
            clap::Arg::new("max_count")
                .long("max-count")
                .short('m')
                .help("Maximum number of matches")
                .default_value("0"),
        )
        .arg(
            clap::Arg::new("include")
                .long("include")
                .help("Glob pattern for file names (e.g. \"*.go\")"),
        )
        .arg(
            clap::Arg::new("max_depth")
                .long("max-depth")
                .short('d')
                .help("Maximum directory depth (0 = unlimited)")
                .default_value("0"),
        )
        .arg(
            clap::Arg::new("json")
                .long("json")
                .action(clap::ArgAction::SetTrue)
                .help("Output in JSON format"),
        )
}

pub async fn run(matches: &Args) -> anyhow::Result<()> {
    let pattern = Some(&matches.pattern).context("pattern required")?;
    let path = Some(&matches.path).map(|s| s.as_str()).unwrap_or(".");
    let ignore_case = matches.ignore_case;
    let max_count: usize = Some(&matches.max_count)
        .and_then(|s| s.parse().ok())
        .unwrap_or(0);
    let include = matches.include.as_deref();
    let max_depth: usize = Some(&matches.max_depth)
        .and_then(|s| s.parse().ok())
        .unwrap_or(0);
    let use_json = matches.json;

    let re_pattern = if ignore_case {
        format!("(?i){pattern}")
    } else {
        pattern.clone()
    };

    let re = regex::Regex::new(&re_pattern).context("invalid regex")?;
    let include_re = include.and_then(include_to_regex);

    let results = search_text_in_root(&re, include_re.as_ref(), path, max_depth, max_count)?;

    if use_json {
        let out = serde_json::json!({
            "pattern": pattern,
            "matches": results.len(),
            "results": results,
        });
        println!("{}", serde_json::to_string_pretty(&out)?);
    } else if results.is_empty() {
        println!("(no matches)");
    } else {
        for m in &results {
            println!("{}:{}: {}", m.file, m.line, m.content);
        }
        if results.len() > 1 {
            println!("\n{} matches", results.len());
        }
    }

    Ok(())
}

fn include_to_regex(include: &str) -> Option<regex::Regex> {
    let glob_re = include.replace('.', "\\.").replace('*', ".*");
    let re_str = format!("^{glob_re}$");
    regex::Regex::new(&re_str).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_include_to_regex_exact() {
        let re = include_to_regex("main.rs").unwrap();
        assert!(re.is_match("main.rs"));
        assert!(!re.is_match("main.go"));
    }

    #[test]
    fn test_include_to_regex_wildcard() {
        let re = include_to_regex("*.rs").unwrap();
        assert!(re.is_match("main.rs"));
        assert!(re.is_match("lib.rs"));
        assert!(!re.is_match("main.go"));
    }

    #[test]
    fn test_include_to_regex_prefix_wildcard() {
        let re = include_to_regex("test_*").unwrap();
        assert!(re.is_match("test_foo.rs"));
        assert!(!re.is_match("foo_test.rs"));
    }

    #[test]
    fn test_include_to_regex_dot_in_pattern() {
        let re = include_to_regex("*.min.js").unwrap();
        assert!(re.is_match("app.min.js"));
        assert!(!re.is_match("app.dev.js"));
    }

    #[test]
    fn test_include_to_regex_invalid_pattern() {
        let re = include_to_regex("[invalid");
        assert!(re.is_none());
    }

    #[test]
    fn test_include_to_regex_empty() {
        let re = include_to_regex("").unwrap();
        assert!(re.is_match(""));
        assert!(!re.is_match("anything"));
    }
}

fn search_text_in_root(
    re: &regex::Regex,
    include: Option<&regex::Regex>,
    root: &str,
    max_depth: usize,
    max_count: usize,
) -> anyhow::Result<Vec<TextMatch>> {
    let mut results = Vec::new();
    let mut seen = std::collections::HashSet::new();

    let root_path = std::path::Path::new(root);
    if !root_path.exists() {
        return Ok(results);
    }

    if root_path.is_file() {
        let matches = search_file_text(root_path, re, max_count);
        for m in matches {
            let key = format!("{}:{}", m.file, m.content);
            if seen.insert(key) {
                results.push(m);
            }
        }
        return Ok(results);
    }

    let mut walker = walkdir::WalkDir::new(root);
    if max_depth > 0 {
        walker = walker.max_depth(max_depth);
    }

    for entry in walker.into_iter().filter_map(|e| e.ok()) {
        if !entry.file_type().is_file() {
            continue;
        }

        if let Some(inc) = include {
            let name = entry.file_name().to_string_lossy();
            if !inc.is_match(&name) {
                continue;
            }
        }

        let matches = search_file_text(entry.path(), re, max_count);
        for m in matches {
            let key = format!("{}:{}", m.file, m.content);
            if seen.insert(key) {
                results.push(m);
            }
        }
    }

    Ok(results)
}

fn search_file_text(path: &std::path::Path, re: &regex::Regex, max_count: usize) -> Vec<TextMatch> {
    let data = match std::fs::read(path) {
        Ok(d) => d,
        Err(_) => return Vec::new(),
    };

    if data.contains(&0) {
        return Vec::new();
    }

    let content = match String::from_utf8(data) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let mut matches = Vec::new();
    let file = path.to_string_lossy().to_string();

    for (i, line) in content.lines().enumerate() {
        if re.is_match(line) {
            matches.push(TextMatch {
                file: file.clone(),
                line: i + 1,
                content: line.to_string(),
            });
            if max_count > 0 && matches.len() >= max_count {
                break;
            }
        }
    }

    matches
}
